### PR TITLE
Switch C files to CC files

### PR DESCRIPTION
**Pull request by binji**
Mostly this involves adding additional casts. Though there are a few
more substantial changes:

* The default method for relocating parser stacks no longer works
  because Bison assumes that C++ values can't be memcpy'd. Ours can, but
  there's no easy way to make the generated code do the right thing, so
  we do it manually
* Removed all uses of WabtBool and replaced with bool
* Renamed all uses of export and mutable -> export_ and mutable_
* Casting an invalid value to an enum triggers ubsan, so we have to be a
  little more careful about when we do it (see binary-reader.c:read_sections())
* It's illegal to forward-declare enums, so we just #include instead.
* Designated initializers are not allowed in g++, so we have to switch
  them to lazily initialized structures instead. Pretty horrible, so it
  will be nice to have a better solution for C++.

### CODE REVIEW

**Comment by dschuff** [comment]
<not sure if troll or not> Should we use `static_cast` instead of C-style casts? If we like google/chromium style I believe the answer is yes, right?

**Comment by dschuff** [comment]
oh man, I prefaced that last post with `<not-sure-if-troll-or-not>` but github scrubbed it....

**Comment by binji** [comment]
> Should we use static_cast instead of C-style casts? If we like google/chromium style I believe the answer is yes, right?

Yes, definitely. I just figured it would make sense to scrub the entire codebase of C-style casts in separate PR.

File: src/ast-parser.y [hunk 1]
```diff
@@ -28,16 +28,26 @@
 
 #define INVALID_VAR_INDEX (-1)
 
-/* the default value for YYMAXDEPTH is 10000, which can be easily hit since our
-   grammar is right-recursive.
+#define RELOCATE_STACK(type, array, stack_base, old_byte_size, new_size)      \
+  do {                                                                        \
+    if ((stack_base) == (array)) {                                            \
+      (stack_base) = (type*)wabt_alloc((new_size) * sizeof(*(stack_base)));   \
+      memcpy((stack_base), (array), old_byte_size);                           \
```
Comment (dschuff) [inline src/ast-parser.y]: Is there somewhere we can put `static_assert`s to ensure that this remains true?

File: src/ast-writer.cc [hunk 2]
```diff
@@ -184,12 +184,12 @@ static void write_string_slice(Context* ctx,
...
 }
 
-static WabtBool write_string_slice_opt(Context* ctx,
+static bool write_string_slice_opt(Context* ctx,
                                        const WabtStringSlice* str,
                                        NextChar next_char) {
   if (str->start)
     write_string_slice(ctx, str, next_char);
-  return str->start ? WABT_TRUE : WABT_FALSE;
+  return str->start != NULL;
```
Comment (dschuff) [inline src/ast-writer.cc]: `nullptr`? Or I guess that's another global PR too.

Comment (jgravelle-google) [inline src/ast-writer.cc]: `#define NULL nullptr` should fix everything

Comment (binji) [inline src/ast-writer.cc]: Yeah, I was thinking in a later PR.

File: src/ast-parser.y (hunk 1 above)
Comment (binji) [inline src/ast-parser.y]: Sorry, ensure that what remains true?

**Comment by binji** [comment]
OK, looks like appveyor is finally happy. Travis should be OK too, PTAL

File: src/ast-parser.y (hunk 1 above)
Comment (dschuff) [inline src/ast-parser.y]: Ensure that memcpy is legal, i.e. assert that `std::is_trivially_copyable<Foo>` or however that bit is spelled.

Comment (binji) [inline src/ast-parser.y]: Ah, got it. You mind if I file a bug and fix in another PR so I can land this while everything is still building? :-)

File: src/config.h.in [hunk 3]
```diff
@@ -64,12 +64,24 @@
...
 #define WABT_UNLIKELY(x) __builtin_expect(!!(x), 0)
 #define WABT_LIKELY(x) __builtin_expect(!!(x), 1)
 #define WABT_PRINTF_FORMAT(format_arg, first_arg) \
   __attribute__((format(printf, (format_arg), (first_arg))))
 
+#ifdef __cplusplus
+#if __cplusplus >= 201103L
+#define WABT_STATIC_ASSERT(x) static_assert((x), #x)
+#else
+#define WABT_STATIC_ASSERT__(x, c) \
```
Comment (sbc100) [inline src/config.h.in]: Is this for < C++11?   What is the min version we are supporting?

File: src/ast-writer.cc (hunk 2 above)
Comment (binji) [inline src/ast-writer.cc]: Added to the checklist in #308

File: src/config.h.in (hunk 3 above)
Comment (binji) [inline src/config.h.in]: Yeah, should support C++11 only. I guess this should be removed.

**Comment by sbc100** [comment]
OMG C++!  Got for it, its the nineties.

**Comment by sbc100** [comment]
Also, can we start using python3 features now :fearful:

**Comment by binji** [comment]
Hey, watch it or we'll switch to haskell.

**Comment by binji** [comment]
(as for python3: never)
(serious answer: nice to support 2 and 3 (currently broken btw))

**Pull request merged**